include/oneapi/mkl/dft/detail/external_workspace_helper.hpp

```diff
@@ -0,0 +1,178 @@
+
+#ifndef _ONEMKL_DFT_EXTERNAL_WORKSPACE_HELPER_HPP_
+#define _ONEMKL_DFT_EXTERNAL_WORKSPACE_HELPER_HPP_
+
+#if __has_include(<sycl/sycl.hpp>)
+#include <sycl/sycl.hpp>
+#else
+#include <CL/sycl.hpp>
+#endif
+
+#include "oneapi/mkl/detail/export.hpp"
+#include "oneapi/mkl/dft/detail/types_impl.hpp"
+#include "oneapi/mkl/dft/detail/commit_impl.hpp"
+
+namespace oneapi {
+namespace mkl {
+namespace dft {
+namespace detail {
+
+template <precision prec, domain dom>
+class external_workspace_helper {
+public:
+    using commit_impl_t = commit_impl<prec, dom>;
+    using scalar_t = typename commit_impl_t::scalar_type;
+
+private:
+    // Enum to represent whatever the workspace was set as.
+    enum class ext_workspace_type {
+        not_set,
+        usm,
+        buffer,
+    };
+
+    // Is an external workspace required?
+    bool m_ext_workspace_rqd;
+
+    // Set workspace type, with optional workspaces.
+    ext_workspace_type m_workspace_type;
+
+    // Minimum size of workspace in bytes. -1 indicates not set.
+    std::int64_t m_workspace_bytes_rqd;
+
+    std::optional<sycl::buffer<scalar_t>> m_workspace_buffer;
+
+public:
+    /** Constructor.
+     *  @param ext_workspace_rqd True if WORKSPACE_PLACEMENT is set to WORKSPACE_EXTERNAL.
+     *  @param backend_rqd_bytes_fn The function to call to obtain the backend's required workspace size.
+    */
+    inline constexpr external_workspace_helper(bool ext_workspace_rqd)
+            : m_ext_workspace_rqd(ext_workspace_rqd),
+              m_workspace_type(ext_workspace_type::not_set),
+              m_workspace_bytes_rqd(-1) {}
+
+    /** Reset. May be required for changing descriptor settings before recommitting.
+     *  @param ext_workspace_rqd True if WORKSPACE_PLACEMENT is set to WORKSPACE_EXTERNAL.
+    */
+    inline void reset(bool ext_workspace_rqd) {
+        *this = external_workspace_helper(ext_workspace_rqd);
+    }
+
+    /** Get the required workspace bytes for the backend's external workspace.
+     *  @param committed_desc The backend's native descriptor.
+    */
+    inline std::int64_t get_rqd_workspace_bytes(commit_impl_t& committed_desc) {
+        if (m_workspace_bytes_rqd == -1) {
+            m_workspace_bytes_rqd = committed_desc.get_workspace_external_bytes_impl();
+        }
+        return m_workspace_bytes_rqd;
+    }
+
+    /** Throw according to spec for setting the workspace. USM version.
+     *  @param committed_desc The backend's native descriptor.
+     *  @param usm_workspace A USM allocation for the workspace. Assumed to be sufficeintly large.
+    */
+    inline void set_workspace_throw(commit_impl_t& committed_desc, scalar_t* usm_workspace) {
+        if (get_rqd_workspace_bytes(committed_desc) > 0 && usm_workspace == nullptr) {
+            throw mkl::invalid_argument("DFT", "set_workspace",
+                                        "Provided workspace is too small (nullptr)");
```

Comment:
I think it would be better to clearly say that the function was expecting a non-null pointer.

Reply:
The backend might not require an external workspace, in which case the user might legally provide a `nullptr`.

Reply:
I'm just commenting on the explanation string. "too small" seems odd to me.
```suggestion
                                        "Backend expected a non-null workspace");
```

Reply:
Done in [1da13ee](https://github.com/oneapi-src/oneMKL/pull/411/commits/1da13eef0fbb663fcadc1abbcc15ad930f977589).